adapters/metax/metax.go

```diff
@@ -0,0 +1,244 @@
+			continue
+		}
+
+		if err := preprocessImp(&impCopy); err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		endpoint, err := a.getEndpoint(metaxExt)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		impCopy.Ext = nil
+		requestCopy.Imp = []openrtb2.Imp{impCopy}
+		reqJSON, err := json.Marshal(requestCopy)
+		if err != nil {
+			errs = append(errs, err)
+			return nil, errs
+		}
+
+		headers := http.Header{}
+		headers.Add("Content-Type", "application/json;charset=utf-8")
+		headers.Add("Accept", "application/json")
+		reqDatas = append(reqDatas, &adapters.RequestData{
+			Method:  "POST",
+			Uri:     endpoint,
+			Body:    reqJSON,
+			Headers: headers,
+			ImpIDs:  openrtb_ext.GetImpIDs(requestCopy.Imp),
+		})
+	}
+
+	return reqDatas, errs
+}
+
+func (a *adapter) MakeBids(bidReq *openrtb2.BidRequest, reqData *adapters.RequestData, respData *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if respData.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if respData.StatusCode == http.StatusBadRequest {
+		return nil, []error{&errortypes.BadInput{
+			Message: fmt.Sprintf("Unexpected status code: %d. Bad request", respData.StatusCode),
+		}}
+	}
+
+	if respData.StatusCode != http.StatusOK {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d", respData.StatusCode),
+		}}
+	}
+
+	var bidResp openrtb2.BidResponse
+	if err := json.Unmarshal(respData.Body, &bidResp); err != nil {
+		return nil, []error{err}
+	}
+
+	// additional no content check
+	if len(bidResp.SeatBid) == 0 || len(bidResp.SeatBid[0].Bid) == 0 {
+		return nil, nil
+	}
+
+	resp := adapters.NewBidderResponseWithBidsCapacity(len(bidResp.SeatBid[0].Bid))
+	for _, sb := range bidResp.SeatBid {
+		for i := range sb.Bid {
+			bid := &sb.Bid[i]
+			resp.Bids = append(resp.Bids, &adapters.TypedBid{
+				Bid:     bid,
+				BidType: getBidType(bidReq.Imp, bid.ImpID),
+			})
+		}
+	}
+	return resp, nil
+}
+
+func (a *adapter) getEndpoint(ext *openrtb_ext.ExtImpMetaX) (string, error) {
+	params := macros.EndpointTemplateParams{
+		PublisherID: url.PathEscape(ext.PublisherID),
+		AdUnit:      url.PathEscape(ext.Adunit),
+	}
+	return macros.ResolveMacros(a.template, params)
+}
+
+func parseBidderExt(imp *openrtb2.Imp) (*openrtb_ext.ExtImpMetaX, error) {
+	var bidderExt adapters.ExtImpBidder
+	if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+		return nil, err
+	}
+
+	var metaxExt openrtb_ext.ExtImpMetaX
+	if err := json.Unmarshal(bidderExt.Bidder, &metaxExt); err != nil {
+		return nil, errors.New("Wrong MetaX bidder ext")
+	}
+
+	return &metaxExt, nil
+}
+
+// publisher ID and adunit is numeric at this moment
+func validateParams(ext *openrtb_ext.ExtImpMetaX) error {
+	if _, err := strconv.Atoi(ext.PublisherID); err != nil {
+		return errors.New("invalid publisher ID")
+	}
+	if _, err := strconv.Atoi(ext.Adunit); err != nil {
+		return errors.New("invalid adunit")
+	}
```

Comment:
I have looked at some other adapters and found that most publisherId and adunit are in string format. However, in our system these two fields are integer type, so we use string type to avoid configuration errors caused by user habits or misunderstanding of the type, and then check in our own adapter whether it is a numeric string

Based on your experience, if it is better to configure integer type, I will adjust it later

Reply:
I have looked at some other adapters and found that most publisherId and adunit are in string format. However, in our system these two fields are integer type, so we use string type to avoid configuration errors caused by user habits or misunderstanding of the type, and then check in our own adapter whether it is a numeric string

Based on your experience, if it is better to configure integer type, I will adjust it later